designs/2024-repo-ecosystem-plugin-tests/README.md

```diff
@@ -0,0 +1,153 @@
+- Repo: eslint/eslint
+- Start Date: 2024-11-25
+- RFC PR: <https://github.com/eslint/rfcs/pull/127>
+- Authors: [Josh Goldberg](https://github.com/JoshuaKGoldberg)
+
+# Introduce ecosystem tests for popular third-party plugins
+
+## Summary
+
+Adding an CI job to the `eslint/eslint` repo that checks changes against a small selection of third-party plugins.
+
+## Motivation
+
+Changes in ESLint occasionally break downstream plugins in unexpected ways.
+Those changes might be unintentional breaking changes, or even non-breaking changes that happen to touch edge case behaviors relied on by plugins.
+
+[Bug: Error while loading rule '@typescript-eslint/no-unused-expressions](https://github.com/eslint/eslint/issues/19134) is an example change in ESLint's that caused downstream breakages in third-party plugins.
+At least two popular plugins -[`eslint-plugin-unicorn`](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/2496) and [`typescript-eslint`](https://github.com/typescript-eslint/typescript-eslint/issues/10338)- were broken by that change.
+
+The plugins broke because they were relying on non-public implementation details of ESLint rules per [Docs: Formalize recommendation against plugins calling to rules via use-at-your-own-risk](https://github.com/eslint/eslint/issues/19169).
+When the root cause is a bug in the downstream plugins, an "early warning" system would help them fix their issues before the incompatible changes to ESLint are published.
+
+## Detailed Design
+
+### CI Job
+
+The new CI job will, for each plugin:
+
+1. Create a new directory containing a `package.json`, `eslint.config.js`, and small set of files known to be parsed and not cause lint reports with the plugin
```

Comment:
Can you clarify exactly what should be in `package.json`, `eslint.config.js`?

some more background:
The eslint ci process is already using (testing) some popular eslint plugins, e.g. `eslint-plugin-unicorn`, `eslint-plugin-eslint-plugin`, `eslint-plugin-jsdoc`...
https://github.com/eslint/eslint/blob/c61870715e5bc1cb51c45a8efd72f392502807b8/packages/eslint-config-eslint/base.js#L268-L270
https://github.com/eslint/eslint/blob/c61870715e5bc1cb51c45a8efd72f392502807b8/packages/eslint-config-eslint/nodejs.js#L12-L31

Unfortunately, it didn't catch this problem (since it was not using `unicorn/expiring-todo-comments`); to avoid this, does this mean we should use as many rules and features as possible in this config?

Reply:
This is a great point, adding to the RFC:

* A note that ESLint core doesn't always enable all rules, even when it uses the plugin
* `eslint-config-eslint` isn't always up-to-date with the latest versions of dependencies

So even if ESLint core were to enable many more rules, that set wouldn't likely catch as many rules as this system, and it wouldn't always catch them on the latest version.